### PR TITLE
feat(services): add aggregations

### LINKED ISSUE #249: feat(service):   Service List View Aggregation Query
**Task Description**

For performance reasons an aggregation of the counts of component instances and issue matches is required.

**Acceptance Criteria:**

- [ ] Count component instances
- [ ] Count issue matches

**Expected Test:**
- Test aggregations in the respective layers

### PR DESCRIPTION
**Pull request by MR2011**
## Description
This PR adds the following two aggregations for services:
- Count of ComponentInstances
- Count of IssueMatches

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Closes #249 

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Tests:
- E2E testing metadata (aggregation)
- App testing ListServices with aggregation
- DB testing individual aggregation functions

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [x] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/database/mariadb/service.go [hunk 1]
```diff
@@ -252,6 +284,59 @@ func (s *SqlDatabase) GetServices(filter *entity.ServiceFilter) ([]entity.Servic
 	)
 }
 
+func (s *SqlDatabase) GetServicesWithComponentInstanceCount(filter *entity.ServiceFilter) ([]entity.ServiceWithAggregations, error) {
```
Comment (drochow) [inline internal/database/mariadb/service.go:287]: Actually, I like having specialized queries for that.

I think we should go 1 step further and do leave away the genralization for "serviceWithAggregation" that we've build previously. 

As this example shows aggregations can easily get wrong when you add additional aggregations and even making combinations of certain aggregated values impossible. 
As such we might should  consider each aggregation separate of each other?

File: internal/app/service/service_handler.go [hunk 2]
```diff
@@ -38,6 +42,62 @@ func NewServiceHandlerError(msg string) *ServiceHandlerError {
...
+	ciIds := maps.Keys(ciCounts)
+	imIds := maps.Keys(imCounts)
+
+	slices.Sort(ciIds)
+	slices.Sort(imIds)
+
+	// check if same services were returned by the aggregation queries
+	if !slices.Equal(ciIds, imIds) {
+		return nil, fmt.Errorf("aggregation queries returned different services")
+	}
```
Comment (drochow) [inline internal/app/service/service_handler.go:83]: Do we really want to check that? 

When a certain service has a 0 count. Is it then still be listed with count = 0 ?

File: internal/app/service/service_handler_test.go [hunk 3]
```diff
@@ -109,6 +111,96 @@ var _ = Describe("When listing Services", Label("app", "ListServices"), func() {
 			Entry("When  pageSize is 10 and the database was returning 11 elements", 10, 11, 10, true),
 		)
 	})
```
Comment (drochow) [inline internal/app/service/service_handler_test.go:113]: If we want to enforce the above with checking that the two counts where giving back results for the same set of services we should also add a test for that

File: internal/database/mariadb/service_test.go [hunk 4]
```diff
@@ -436,6 +436,61 @@ var _ = Describe("Service", Label("database", "Service"), func() {
...
+				//Should be filled with a check for each aggregation value,
+				// this is currently skipped due to the complexity of the test implementation
+				// as we would need to implement for each of the aggregations a manual aggregation
+				// based on the seederCollection.
+				//
+				// This tests should therefore only get implemented in case we encourage errors in this area to test against
+				// possible regressions
+			})
+		})
+	})
```
Comment (drochow) [inline internal/database/mariadb/service_test.go:493]: Don't we need a test then here that is checking that for 0 counts the service is still listed? 

So I'm expecting 10 services even though they might not all have a count > 0.

**Review by drochow — CHANGES_REQUESTED** [review]

**Review by drochow — CHANGES_REQUESTED** [review]
I still think we might can create 2 services (without anything) and we would expect them to come back no?

**Comment by drochow** [comment]
@MR2011  Am I missing something? I think its just about calling: 

```golang
seeder.SeedServices(10)
```

in a test and check then if also the 0 count services are returned (if no filter is applied)

**Comment by MR2011** [comment]
> @MR2011 Am I missing something? I think its just about calling:
> 
> ```go
> seeder.SeedServices(10)
> ```
> 
> in a test and check then if also the 0 count services are returned (if no filter is applied)
That's tested here
https://github.com/cloudoperators/heureka/blob/963961c5fefa7a3825a4802b56f43af62c09bf59/internal/database/mariadb/service_test.go#L440-L465

**Pull request merged**